### PR TITLE
test: fix data race using sync.Map for the mock backend

### PR DESCRIPTION
**Pull request by mydoomfr**
We're sometimes getting goroutine stack trace errors related to multiple access read/write on the mocked map `zones` and `rrsets` variables.

Switching to a sync.Map is an easy way to benefit from mutex and prevent multiple data race conditions.

The PR also include the remove of all the time.Sleep workaround and try to be fully async in the test cases.

### CODE REVIEW

File: internal/controller/suite_test.go [hunk 1]
```diff
@@ -314,14 +384,22 @@ func getMockedRecordsForType(rrsetName, rrsetType string) (result []string) {
 }
 
 func getMockedTTL(rrsetName, rrsetType string) (result uint32) {
-	rrset := records[makeCanonical(rrsetName)]
+	rrset, ok := readFromRecordsMap(makeCanonical(rrsetName))
+	if !ok || rrset == nil {
+		return
+	}
+
```
Comment (antrema) [inline internal/controller/suite_test.go:391]: See previous comment

File: internal/controller/suite_test.go [hunk 2]
```diff
@@ -314,14 +384,22 @@ func getMockedRecordsForType(rrsetName, rrsetType string) (result []string) {
...
 	}
 	return
 }
 func getMockedComment(rrsetName, rrsetType string) (result string) {
-	rrset := records[makeCanonical(rrsetName)]
+	rrset, ok := readFromRecordsMap(makeCanonical(rrsetName))
+	if !ok || rrset == nil {
+		return
+	}
+
```
Comment (antrema) [inline internal/controller/suite_test.go:402]: See previous comment

File: internal/controller/suite_test.go [hunk 3]
```diff
@@ -295,15 +357,23 @@ func getMockedNameservers(zoneName string) (result []string) {
 }
 
 func getMockedKind(zoneName string) (result string) {
-	zone := zones[makeCanonical(zoneName)]
+	zone, ok := readFromZonesMap(makeCanonical(zoneName))
+	if !ok || zone == nil {
+		return
+	}
+
```
Comment (antrema) [inline internal/controller/suite_test.go:364]: See previous comment

File: internal/controller/suite_test.go [hunk 4]
```diff
@@ -295,15 +357,23 @@ func getMockedNameservers(zoneName string) (result []string) {
...
 	return
 }
 
 func getMockedRecordsForType(rrsetName, rrsetType string) (result []string) {
-	rrset := records[makeCanonical(rrsetName)]
+	rrset, ok := readFromRecordsMap(makeCanonical(rrsetName))
+	if !ok || rrset == nil {
+		return
+	}
+
```
Comment (antrema) [inline internal/controller/suite_test.go:376]: See previous comment

File: internal/controller/suite_test.go [hunk 5]
```diff
@@ -99,14 +159,12 @@ var _ = BeforeSuite(func() {
...
-	zones = map[string]*powerdns.Zone{}
-	records = map[string]*powerdns.RRset{}
 	err = (&RRsetReconciler{
 		Client: k8sManager.GetClient(),
 		Scheme: k8sManager.GetScheme(),
 		PDNSClient: PdnsClienter{
-			Records: m.Records,
-			Zones:   m.Zones,
+			Records: &m.Records,
+			Zones:   &m.Zones,
```
[the remark this answers is not in the document]
Comment (mydoomfr) [reply internal/controller/suite_test.go:167]: Good catch! Fixed

**Review by antrema — APPROVED** [review]

**Pull request merged**